### PR TITLE
Physics-informed neural networks for data-driven fluid model

### PR DESCRIPTION
**Pull request by EvertBunschoten**
## Proposed Changes
The data-driven fluid model in SU2 has been upgraded with the ability to use physics-informed neural networks (PINN's) to compute the fluid thermodynamic state for NICFD flows. This functionality was presented during the 2024 SU2 conference and the issues regarding non-physical behavior have been addressed. 

The use of PINN makes the data-driven fluid model substantially more robust and efficient for flow and adjoint simulations of flows of fluids in the non-ideal thermodynamic state.

PINN's can be trained using [SU2 DataMiner](https://github.com/EvertBunschoten/SU2_DataMiner.git) for any fluid or compatible mixture of fluids available in the CoolProp library. An elaborate tutorial integrating SU2 DataMiner with the data-driven fluid solver in SU2 will follow soon!

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [X] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [X] I used the pre-commit hook to prevent dirty commits and used `pre-commit run --all` to format old commits.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp [hunk 1]
```diff
@@ -137,44 +178,61 @@ void CDataDrivenFluid::MapInputs_to_Outputs() {
...
+  AD::SetPreaccIn(rho);
+  AD::SetPreaccIn(e);
 
-  /*--- Compute speed of sound. ---*/
-  auto blue_term = (dsdrho_e * (2 - rho * pow(dsde_rho, -1) * d2sdedrho) + rho * d2sdrho2);
-  auto green_term = (-pow(dsde_rho, -1) * d2sde2 * dsdrho_e + d2sdedrho);
+  /*--- Compute thermodynamic state based on density and energy. ---*/
+  
+  Density = max(min(rho, rho_max), rho_min);
+  StaticEnergy = max(min(e, e_max), e_min);
```
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CDataDrivenFluid.cpp:189]: These are preacc outs

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp [hunk 2]
```diff
@@ -273,6 +339,9 @@ void CDataDrivenFluid::Run_Newton_Solver(su2double Y1_target, su2double Y2_targe
   /*--- 2D Newton solver, computing the density and internal energy values corresponding to Y1_target and Y2_target.
    * ---*/
 
+  AD::StartPreacc();
+  AD::SetPreaccIn(*Y1);
+  AD::SetPreaccIn(*Y2);
```
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CDataDrivenFluid.cpp:344]: Why are these variables passed by pointer if they are just scalars?

File: SU2_CFD/src/iteration/CFluidIteration.cpp [hunk 3]
```diff
@@ -306,7 +312,7 @@ void CFluidIteration::TurboMonitor(CGeometry**** geometry_container, CConfig** c
           case RIEMANN_BOUNDARY:
             KindBCOption = config->GetKind_Data_Riemann(Marker_Tag);
             if (KindBCOption == STATIC_PRESSURE || KindBCOption == RADIAL_EQUILIBRIUM) {
-              SU2_MPI::Error("Outlet pressure ramp only implemented for NRBC", CURRENT_FUNCTION);
+              config->SetRiemann_Var1(outPres, Marker_Tag);
```
Comment (pcarruscag) [inline SU2_CFD/src/iteration/CFluidIteration.cpp:315]: @joshkellyjak

File: SU2_CFD/src/output/CFlowCompOutput.cpp [hunk 4]
```diff
@@ -250,6 +250,7 @@ void CFlowCompOutput::SetVolumeOutputFields(CConfig *config){
     AddVolumeOutput("EXTRAPOLATION", "Extrapolation", "PRIMITIVE", "Density, energy outside data range");
     AddVolumeOutput("FLUIDMODEL_NEWTONITER", "nIter_Newton", "PRIMITIVE", "Number of iterations evaluated by the Newton solver");
     AddVolumeOutput("ENTROPY", "Entropy", "PRIMITIVE", "Fluid entropy value");
+    AddVolumeOutput("HEAT_CAPACITY", "Heat_Capacity", "PRIMITIVE", "Heat capacity");
```
Comment (pcarruscag) [inline SU2_CFD/src/output/CFlowCompOutput.cpp:253]: Isn't this in another PR already?

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp (hunk 2 above)
Comment (EvertBunschoten) [reply SU2_CFD/src/fluid/CDataDrivenFluid.cpp]: Good point, I changed them to pass by reference.

File: SU2_CFD/src/iteration/CFluidIteration.cpp (hunk 3 above)
Comment (joshkellyjak) [reply SU2_CFD/src/iteration/CFluidIteration.cpp:315]: @EvertBunschoten I made some modifcations to this part of the code in #2419 , I'll try close it today.

Comment (joshkellyjak) [reply SU2_CFD/src/iteration/CFluidIteration.cpp:315]: update: finished now.

File: Common/include/CConfig.hpp [hunk 5]
```diff
@@ -3993,6 +3997,20 @@ class CConfig {
...
+   * \brief Get initial density value for Newton solver in data-driven fluid model.
+   * \return Initial density value.
+   */
+  su2double GetInitialDensity_DataDriven(void) const { return DataDriven_rho_init; }
+
+  /*!
+   * \brief Get initial static value for Newton solver in data-driven fluid model.
+   * \return Initial static value.
+   */
+  su2double GetInitialEnergy_DataDriven(void) const { return DataDriven_e_init; }
```
Comment (pcarruscag) [inline Common/include/CConfig.hpp:4010]: Please make a dedicated struct to store the _DataDriven config options and replace all _DataDriven accessors which are becoming numerous.

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp [hunk 6]
```diff
@@ -137,44 +178,61 @@ void CDataDrivenFluid::MapInputs_to_Outputs() {
...
+  Cv = 1 / dTde_rho;
   dhdrho_P = dhdrho_e - dhde_rho * (1 / dPde_rho) * dPdrho_e;
   dhdP_rho = dhde_rho * (1 / dPde_rho);
   dsdrho_P = dsdrho_e - dPdrho_e * (1 / dPde_rho) * dsde_rho;
   dsdP_rho = dsde_rho / dPde_rho;
+
+  su2double drhode_p = -dPde_rho/dPdrho_e;
+  su2double dTde_p = dTde_rho + dTdrho_e*drhode_p;
+  su2double dhde_p = dhde_rho + drhode_p*dhdrho_e;
+  Cp = dhde_p / dTde_p;
```
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CDataDrivenFluid.cpp:226]: Are Cv and Cp local variables? Why are they not preaccumulated? Do you have any adjoint verification cases yet to be playing with preaccumulation?

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp [hunk 7]
```diff
@@ -137,44 +178,61 @@ void CDataDrivenFluid::MapInputs_to_Outputs() {
...
+  dhde_rho = 1 + dPde_rho / Density;
 
+  /*--- Compute specific heat at constant volume and specific heat at constant pressure. ---*/
+  Cv = 1 / dTde_rho;
   dhdrho_P = dhdrho_e - dhde_rho * (1 / dPde_rho) * dPdrho_e;
   dhdP_rho = dhde_rho * (1 / dPde_rho);
   dsdrho_P = dsdrho_e - dPdrho_e * (1 / dPde_rho) * dsde_rho;
   dsdP_rho = dsde_rho / dPde_rho;
+
+  su2double drhode_p = -dPde_rho/dPdrho_e;
```
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CDataDrivenFluid.cpp:223]: other places too
```suggestion
  const su2double drhode_p = -dPde_rho/dPdrho_e;
```

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp [hunk 8]
```diff
@@ -268,11 +334,14 @@ void CDataDrivenFluid::Evaluate_Dataset(su2double rho, su2double e) {
   }
 }
 
-void CDataDrivenFluid::Run_Newton_Solver(su2double Y1_target, su2double Y2_target, su2double* Y1, su2double* Y2,
-                                         su2double* dY1drho, su2double* dY1de, su2double* dY2drho, su2double* dY2de) {
+void CDataDrivenFluid::Run_Newton_Solver(su2double Y1_target, su2double Y2_target, su2double& Y1, su2double& Y2,
+                                         su2double& dY1drho, su2double& dY1de, su2double& dY2drho, su2double& dY2de) {
```
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CDataDrivenFluid.cpp:338]: Why are some of these passed by value and others by non-const reference? Are they outputs?
Why aren't they preaccumulated either as input or output?

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp (hunk 6 above)
Comment (EvertBunschoten) [reply SU2_CFD/src/fluid/CDataDrivenFluid.cpp:227]: I initially copied the preaccumulation from the Peng-Robinson fluid model. I added some unit tests for fluid model consistency and I'm planning to add some AD unit tests as well.

File: Common/include/CConfig.hpp (hunk 5 above)
Comment (EvertBunschoten) [reply Common/include/CConfig.hpp]: I've replaced the accessors for the CDataDrivenFluid fluid model now with structs. I did a similar thing for the CFluidFlamelet fluid model.

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp (hunk 8 above)
Comment (EvertBunschoten) [reply SU2_CFD/src/fluid/CDataDrivenFluid.cpp]: They could indeed be passed by const value and const reference, since none of them are outputs. I still need to verify whether I did the preaccumulation correctly.

File: Common/include/CConfig.hpp [hunk 9]
```diff
@@ -9939,4 +9818,17 @@ class CConfig {
    */
   LM_ParsedOptions GetLMParsedOptions() const { return lmParsedOptions; }
 
+
+  /*!
+   * \brief Get parsed option data structure for data-driven fluid model.
+   * \return option data structure for data-driven fluid model.
+   */
+  DataDrivenFluid_ParsedOptions GetDataDrivenParsedOptions() const { return datadriven_ParsedOptions; }
```
Comment (pcarruscag) [inline Common/include/CConfig.hpp:9826]:
```suggestion
  const DataDrivenFluid_ParsedOptions& GetDataDrivenParsedOptions() const { return datadriven_ParsedOptions; }
```

File: Common/include/CConfig.hpp [hunk 10]
```diff
@@ -6149,16 +6028,16 @@ class CConfig {
...
-   * \brief Set the angular velocity of the mesh about the z-axis.
-   * \param[in] newRotation_Rate_Z - new rotation rate after computing the ramp value.
-   */
-  void SetRotation_Rate_Z(su2double newRotation_Rate_Z);
+  su2double GetFinalValue(RAMP_TYPE ramp_flag) {
+    if (ramp_flag == RAMP_TYPE::GRID && RampRotatingFrame) return FinalRotation_Rate_Z;
+    else if (ramp_flag == RAMP_TYPE::GRID && RampTranslationFrame) return FinalTranslation_Rate_Y;
+    else if (ramp_flag == RAMP_TYPE::BOUNDARY && RampOutletPressure) return FinalOutletPressure;
+    else if (ramp_flag == RAMP_TYPE::BOUNDARY && RampOutletMassFlow) return FinalOutletMassFlow;
+    else return 0.0;
```
Comment (pcarruscag) [inline Common/include/CConfig.hpp:6039]: I think you merged Josh's branch into yours...

File: SU2_CFD/src/output/CAdjFlowOutput.cpp [hunk 11]
```diff
@@ -351,16 +357,16 @@ void CAdjFlowOutput::LoadVolumeDataAdjScalar(const CConfig* config, const CSolve
   }
 
   if (config->GetKind_Species_Model() == SPECIES_MODEL::FLAMELET) {
-
-    for (auto iCV=0u; iCV < config->GetNControlVars(); iCV++) {
-      const auto& cv_name = config->GetControllingVariableName(iCV);
+    const auto flamelet_config_options = config->GetFlameletParsedOptions();
```
Comment (pcarruscag) [inline SU2_CFD/src/output/CAdjFlowOutput.cpp:360]:
```suggestion
    const auto& flamelet_config_options = config->GetFlameletParsedOptions();
```

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp [hunk 12]
```diff
@@ -268,37 +338,37 @@ void CDataDrivenFluid::Evaluate_Dataset(su2double rho, su2double e) {
   }
 }
 
-void CDataDrivenFluid::Run_Newton_Solver(su2double Y1_target, su2double Y2_target, su2double* Y1, su2double* Y2,
-                                         su2double* dY1drho, su2double* dY1de, su2double* dY2drho, su2double* dY2de) {
+void CDataDrivenFluid::Run_Newton_Solver(const su2double Y1_target, const su2double Y2_target, su2double const & Y1, su2double const & Y2,
+                                         su2double const & dY1drho, su2double const & dY1de, su2double const & dY2drho, su2double const & dY2de) {
```
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CDataDrivenFluid.cpp:342]: etc.
```suggestion
                                         const su2double& dY1drho, su2double const & dY1de, su2double const & dY2drho, su2double const & dY2de) {
```

**Review by pcarruscag — COMMENTED** [review]
Thanks your changes look good, but having the Ramp PR here doesn't.
Probably better to wait for @joshkellyjak to finish the Ramp PR and merge it instead of trying to fix it here.
Check the code QL warnings and const& comments in the meantime.

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp (hunk 12 above)
Comment (EvertBunschoten) [reply SU2_CFD/src/fluid/CDataDrivenFluid.cpp:342]: The terms dY1drho, dY2drho, etc should remain modifiable. That's why they are defined as constant references rather than references to constants.

File: Common/include/CConfig.hpp (hunk 10 above)
Comment (EvertBunschoten) [reply Common/include/CConfig.hpp]: Ah yes that's right...

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp (hunk 12 above)
Comment (pcarruscag) [reply SU2_CFD/src/fluid/CDataDrivenFluid.cpp:342]: That's not the way it works

File: config_template.cfg [hunk 13]
```diff
@@ -735,19 +735,20 @@ MIXINGPLANE_INTERFACE_KIND= LINEAR_INTERPOLATION
...
+% Parameters of the outlet Giles ramp (starting outlet value,
 % updating-iteration-frequency, total number of iteration for the ramp)
-RAMP_OUTLET_PRESSURE_COEFF= (400000.0, 10.0, 500)
+RAMP_OUTLET_COEFF= (400000.0, 10.0, 500)
 %
-% Specify ramp option for rotating frame (YES, NO) default NO
-RAMP_ROTATING_FRAME= YES
+% Specify ramp option for frame of motion (YES, NO) default NO
+RAMP_MOTION_FRAME= YES
 %
```
Comment (pcarruscag) [inline config_template.cfg:747]: What's happening here?

File: Common/include/CConfig.hpp [hunk 14]
```diff
@@ -535,9 +535,15 @@ class CConfig {
...
 
-  su2double DataDriven_Relaxation_Factor; /*!< \brief Relaxation factor for Newton solvers in data-driven fluid models. */
+  DataDrivenFluid_ParsedOptions datadriven_ParsedOptions;
+  //ENUM_DATADRIVEN_METHOD Kind_DataDriven_Method;       /*!< \brief Method used for datset regression in data-driven fluid models. */
+
+  //su2double DataDriven_Relaxation_Factor; /*!< \brief Relaxation factor for Newton solvers in data-driven fluid models. */
+  // su2double DataDriven_rho_init{-1.0},    /*!< \brief Optional initial density value for Newton solver in data-driven fluid model.*/
+  //           DataDriven_e_init{-1.0};      /*!< \brief Optional initial static energy value for Newton solver in data-driven fluid model.*/
+
+  //bool UsePINN = false;       /*!< \brief Use physics-informed method for data-driven fluid modeling. */
```
Comment (pcarruscag) [inline Common/include/CConfig.hpp:546]: remove

File: Common/include/geometry/CGeometry.hpp [hunk 15]
```diff
@@ -954,6 +954,19 @@ class CGeometry {
...
+   * \param[in] ramp_flag - flag for type of ramp
+   * \param[in] print - bool to print update to screen
+   */
+  void SetVelocity(CConfig* config, bool print) {
+    if (config->GetKind_GridMovement() == ENUM_GRIDMOVEMENT::ROTATING_FRAME)
+      SetRotationalVelocity(config, print);
+    else if (config->GetKind_GridMovement() == ENUM_GRIDMOVEMENT::STEADY_TRANSLATION)
+      SetTranslationalVelocity(config, print);
+  };
+
```
Comment (pcarruscag) [inline Common/include/geometry/CGeometry.hpp:969]: used?

Comment (EvertBunschoten) [reply Common/include/geometry/CGeometry.hpp]: No, this was still left from the merge of Josh' PR. I fixed it now

**Comment by pcarruscag** [comment]
LGTM but @joshkellyjak should approve because I expect the way the merge and unmerge with the turbo ramp branch was done is going to create either conflicts or silently revert the work in that branch.

**Comment by joshkellyjak** [comment]
> LGTM but @joshkellyjak should approve because I expect the way the merge and unmerge with the turbo ramp branch was done is going to create either conflicts or silently revert the work in that branch.

I'm happy to approve this and then update in turbo ramps branch. Still having the same regression test issue and couldn't find a solution. Probably easier to merge this and then fix on my side.

**Review by joshkellyjak — APPROVED** [review]

File: SU2_CFD/src/fluid/CDataDrivenFluid.cpp (hunk 6 above)
Comment (EvertBunschoten) [reply SU2_CFD/src/fluid/CDataDrivenFluid.cpp:230]: I ran a gradient validation study where I found the maximum difference between the AD gradients and FD gradients to be 0.14%. @joshkellyjak and I did a preaccumulation debugging session together where we identified several instances in the CDataDrivenFluid class where the preaccumulation tag was wrong. After fixing those, I ran the AD again, but the percentage difference in AD gradients was in the order of 1e-5.
